cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -377,12 +384,19 @@ func (m *McmManager) GetMachineDeploymentNodes(machinedeployment *MachineDeploym
 	// Bearing O(n2) complexity, assuming we will not have lot of nodes/machines, open for optimisations.
 	for _, machine := range machinelist.Items {
 		if strings.Contains(machine.Name, md.Name) {
+			var found bool
 			for _, node := range nodelist.Items {
 				if machine.Labels["node"] == node.Name {
 					nodes = append(nodes, node.Spec.ProviderID)
+					found = true;
 					break
 				}
 			}
+			if !found {
+				// No node found - either the machine has not registered yet or AWS is unable to fufill the request.
+				// Report a special ID so that the autoscaler can track it as an unregistered node.
+				nodes = append(nodes, fmt.Sprintf("requested://%s", machine.Name))
```

Comment:
Do you have more insights on, how does autoscaler treat the unregistered nodes, I hope it doesnt care about the exact name of the future node?

Reply:
It uses the node name in a few places as far as I can tell:
1. Uses the node name to determine if it is unregistered by comparing the providerID with the node ID.
2. Tracks how long the node ID has been unregistered for.  It maintains a map of unregistered nodes, along with how long it has been unregistered for.
3. Keeps track of overall total of unregistered nodes, it uses this to fix up the node pool if it takes too long to scale up (by reducing the target size).

I believe that's it - so the important thing is returning a consistent ID, and also returning something to track the unregistered node.

Only issue I found is that MCM will constantly delete / create a new node when InvalidInstanceCapacity is hit, which means these IDs are short lived. Fortunately autoscaler still handles this with the fix up. It would probably be better though if MCM handled this more gracefully.

Reply:
Thanks. 
MCM part should be taken care of by this PR: https://github.com/gardener/machine-controller-manager/pull/525.
- We introduced a new phase called `CrashLoopBackOff` . This is set when machine-creation fails, and now the machine-object is not removed, rather it only retries after a constant delay. Exponential-delay in the future maybe.

Reply:
>I believe that's it - so the important thing is returning a consistent ID, and also returning something to track the unregistered node.

Sure, we are good to go then.